[mlir][Vector] Update VectorEmulateNarrowType.cpp (2/N)

banach-space opened the pull request:
This is PR 2 in a series of N patches aimed at improving
"VectorEmulateNarrowType.cpp". This is mainly minor refactoring, no
major functional changes are made/added.

**CHANGE 1** 

Renames the variable "scale". Note, "scale" could mean either:

  * "container-elements-per-emulated-type", or
  * "emulated-elements-per-container-type".

While from the context it is clear that it's always the former (original
type is always a sub-byte type and the emulated type is usually `i8`),
this PR reduces the cognitive load by making this clear.

**CHANGE 2** 

Replaces `isUnalignedEmulation` with `isFullyAligned`

Note, `isUnalignedEmulation` is always computed following a
"per-element-alignment" condition:
```cpp
// Check per-element alignment.
if (containerBits % emulatedBits != 0) {
  return rewriter.notifyMatchFailure(
    op, "impossible to pack emulated elements into container elements "
    "(bit-wise misalignment)");
}

// (...)

bool isUnalignedEmulation = origElements % emulatedPerContainerElem != 0;
```

Given that `isUnalignedEmulation` captures only one of two conditions
required for "full alignment", it should be re-named as
`isPartiallyUnalignedEmulation`. Instead, I've flipped the condition and
renamed it as `isFullyAligned`:

```cpp
bool isFullyAligned = origElements % emulatedPerContainerElem == 0;
```

**CHANGE 3**
  * Unifies various comments throughout the file (for consistency).
  * Adds new comments throughout the file and adds TODOs where high-level
    comments are missing.
    
    
**GitHub issue to track this work**: https://github.com/llvm/llvm-project/issues/123630

banach-space committed:
[mlir][Vector] Update VectorEmulateNarrowType.cpp (3/N)

This is PR 3 in a series of N patches aimed at improving
"VectorEmulateNarrowType.cpp". This is mainly minor refactoring, no
major functional changes are made/added.

1. Replaces `isUnalignedEmulation` with `isFullyAligned`

Note, `isUnalignedEmulation` is always computed following a
"per-element-alignment" condition:
```cpp
// Check per-element alignment.
if (containerBits % emulatedBits != 0) {
  return rewriter.notifyMatchFailure(
    op, "impossible to pack emulated elements into container elements "
    "(bit-wise misalignment)");
}

// (...)

bool isUnalignedEmulation = origElements % emulatedPerContainerElem != 0;
```

Given that `isUnalignedEmulation` captures only one of two conditions
required for "full alignment", it should be re-named as
`isPartiallyUnalignedEmulation`. Instead, I've flipped the condition and
renamed it as `isFullyAligned`:

```cpp
bool isFullyAligned = origElements % emulatedPerContainerElem == 0;
```

2. In addition:
  * Unifies various comments throughout the file (for consistency).
  * Adds new comments throughout the file and adds TODOs where high-level
    comments are missing.

dcaballe reviewed (commented):
This is continuing the renaming, right? Could we merge with the previous PR? It's difficult to see what the final picture is

lialan commented:
May I propose we add a comment section at the beginning of the code that explains the naming conventions? this will eliminate most of the future naming issues and avoids back-and-forth reviews about names.

banach-space commented:
>  Could we merge with the previous PR? 

Currently as a draft:
* https://github.com/llvm/llvm-project/pull/123633

> It's difficult to see what the final picture is

Trying to capture it here:
* https://github.com/llvm/llvm-project/issues/123630

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -282,13 +282,15 @@ emulatedVectorLoad(OpBuilder &rewriter, Location loc, Value base,
                    Type emulatedElemType) {
-  auto scale = emulatedElemType.getIntOrFloatBitWidth() /
-               origElemType.getIntOrFloatBitWidth();
+  auto elementsPerContainerType = emulatedElemType.getIntOrFloatBitWidth() /
```
`elemPerContainerType` to make it a little bit shorter?

banach-space replied:
Thanks for the suggestion! I'd like to introduce something even less ambiguous, so let me propose:
* `emulatedPerContainerElem`.

I'll send an update shortly and you can tell me what you think :)

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -282,13 +282,15 @@ emulatedVectorLoad(OpBuilder &rewriter, Location loc, Value base,
   return rewriter.create<vector::BitCastOp>(
-      loc, VectorType::get(numEmultedElementsToLoad * scale, origElemType),
+      loc,
+      VectorType::get(numEmultedElementsToLoad * elementsPerContainerType,
```
typo in `numEmultedElementsToLoad` ?

banach-space replied:
Thanks! This an other typos in `emulatedVectorLoad` are fixed in:
* https://github.com/llvm/llvm-project/pull/125415

lialan reviewed (approved):
just some suggestions on the naming.

pashu123 commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -314,14 +316,14 @@ struct ConvertVectorStore final : OpConversionPattern<vector::StoreOp> {
+    int oldBits = oldElementType.getIntOrFloatBitWidth();
+    int newBits = newElementType.getIntOrFloatBitWidth();
 
-    if (dstBits % srcBits != 0) {
```
The naming of srcBits and oldBits is quite confusing. Could we use a more descriptive term, like emulatedBits, instead?

banach-space replied:
https://github.com/llvm/llvm-project/pull/123526 ;-)

pashu123 commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -314,14 +316,14 @@ struct ConvertVectorStore final : OpConversionPattern<vector::StoreOp> {
-          op, "only dstBits % srcBits == 0 supported");
+    // Check per-element alignment.
+    if (newBits % oldBits != 0) {
+      return rewriter.notifyMatchFailure(op, "unalagined element types");
```
```suggestion
      return rewriter.notifyMatchFailure(op, "unaligned element types");
```

banach-space replied:
Addressed in https://github.com/llvm/llvm-project/pull/123526

pashu123 commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -314,14 +316,14 @@ struct ConvertVectorStore final : OpConversionPattern<vector::StoreOp> {
+      return rewriter.notifyMatchFailure(op, "unalagined element types");
     }
-    int scale = dstBits / srcBits;
+    int elementsPerContainerType = newBits / oldBits;
```
is it `elementsPerContainer`? It doesn't seem like type.

banach-space replied:
I've just switched to `emulatedPerContainerElem` - WDYT?

pashu123 commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -385,17 +387,17 @@ struct ConvertVectorMaskedStore final
-          op, "only dstBits % srcBits == 0 supported");
+    // Check per-element alignment.
+    if (newBits % oldBits != 0) {
+      return rewriter.notifyMatchFailure(op, "unalagined element types");
```
```suggestion
      return rewriter.notifyMatchFailure(op, "unaligned element types");
```

banach-space replied:
Addressed in https://github.com/llvm/llvm-project/pull/123526

pashu123 commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -493,14 +497,14 @@ struct ConvertVectorLoad final : OpConversionPattern<vector::LoadOp> {
-          op, "only dstBits % srcBits == 0 supported");
+    // Check per-element alignment.
+    if (newBits % oldBits != 0) {
+      return rewriter.notifyMatchFailure(op, "unalagined element types");
```
```suggestion
      return rewriter.notifyMatchFailure(op, "unaligned element types");
```

pashu123 commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -596,14 +601,14 @@ struct ConvertVectorMaskedLoad final
-          op, "only dstBits % srcBits == 0 supported");
+    // Check per-element alignment.
+    if (newBits % oldBits != 0) {
+      return rewriter.notifyMatchFailure(op, "unalagined element types");
```
```suggestion
      return rewriter.notifyMatchFailure(op, "unaligned element types");
```

banach-space replied:
Addressed in https://github.com/llvm/llvm-project/pull/123526

pashu123 reviewed (commented):
nits

banach-space committed:
[mlir][Vector] Update VectorEmulateNarrowType.cpp (2/N)

This is PR 2 in a series of N patches aimed at improving
"VectorEmulateNarrowType.cpp". This is mainly minor refactoring, no
major functional changes are made/added.

This PR renames the variable "scale". Note, "scale" could mean either:

  * "original-elements-per-emulated-type", or
  * "emulated-elements-per-original-type".

While from the context it is clear that it's always the former (original
type is always a sub-byte type and the emulated type is usually `i8`),
this PR reduces the cognitive load by making this clear.

**DEPENDS ON:**
* #123526 123526

Please only review the [top
commit](https://github.com/llvm/llvm-project/pull/123527/commits/d40b31bb098e874be488182050c68b887e8d091a).

**GitHub issue to track this work**:
https://github.com/llvm/llvm-project/issues/123630

banach-space commented:
@pashu123 , thanks for the review. Looks you have reviewed changes from one of the dependencies of this PR:
* https://github.com/llvm/llvm-project/pull/123526

:) Note that that has been merged and those changes are no longer present in this PR (and all of your suggestions have been incorporated 🙏🏻 ). 

I always list dependencies in the summary - suggestions on how to make this clearer are very welcome. Perhaps it's time to try:
* https://ejoffe.github.io/spr/

banach-space commented:
**UPDATE (3/2/24)**: All dependencies of this PR have been merged + rebased on top of `main`.

dcaballe reviewed (approved):
Thanks for the improvements! LGTM. Let's try to concentrate all the renaming, typo fixing, etc. to this file within a single PR. It makes hard to get a clear picture of the final state.

banach-space commented:
> Let's try to concentrate all the renaming, typo fixing, etc. to this file within a single PR. It makes hard to get a clear picture of the final state.

IIUC, you are suggesting that I include [PATCH 3/N](https://github.com/llvm/llvm-project/pull/123528/commits/b8732cc9820b85ef2b6881bf982ae3dd40a53356) in this PR? That's fine with me, but I'd also like to hear from @lialan and @pashu123 as two other active reviewers. 

Personally, I prefer smaller, isolated changes (IMHO, GitHub UI is pretty bad for bigger changes). However, ultimately, my priority is to make this easy to review.

So, kind request for further feedback:
* 🚢  - merge this as is, i.e. ship it! (*)
* 👍🏻 - I am OK with [PATCH 3/N](https://github.com/llvm/llvm-project/pull/123528/commits/b8732cc9820b85ef2b6881bf982ae3dd40a53356) being included here (and no need for further review is required).

(*) i.e. leave [PATCH 3/N](https://github.com/llvm/llvm-project/pull/123528/commits/b8732cc9820b85ef2b6881bf982ae3dd40a53356) within #123528

hanhanW commented:
I'm +1 on having them in a single PR because the change is not big, and they all touch the same file. Having a clear PR description is good enough to me. IMHO, it also makes the codebase state and commits tracking easier. (2) + (3) is like +107 -78 lines, which is a reasonable size of PR to me.

lialan commented:
> I'm +1 on having them in a single PR because the change is not big, and they all touch the same file. Having a clear PR description is good enough to me. IMHO, it also makes the codebase state and commits tracking easier. (2) + (3) is like +107 -78 lines, which is a reasonable size of PR to me.

+1 too.

banach-space commented:
Thanks for the feedback! Merged https://github.com/llvm/llvm-project/pull/123528 into this PR.

banach-space commented:
I will wait ~24hrs. If there's no new comments, I will assume that folks are happy with these changes and merge it.
